### PR TITLE
Added SetId functionality

### PR DESCRIPTION
**Pull request by AliRKat**

### CODE REVIEW

**Comment by AliRKat** [comment]
[[Unity] implement "setID"](https://trello.com/c/r9lENtX9/1611-unity-implement-setid)

File: Assets/Tests/PlayModeTests/DeviceIdTests.cs [hunk 1]
```diff
@@ -377,22 +379,100 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
...
+        public void SetId_OverDeveloperProvided(string id, DeviceIdType expectedIdType, bool isExpectingChange)
+        {
+            Countly.Instance.Init(TestUtility.CreateBaseConfig());
+            string deviceID = Countly.Instance.Device.DeviceId; // "test_user"
+
+            ValidateDeviceIDAndType(Countly.Instance, deviceID, DeviceIdType.DeveloperProvided, true);
+            Countly.Instance.Device.SetId(id);
+            if (isExpectingChange)
+            {
+                ValidateDeviceIDAndType(Countly.Instance, id, expectedIdType, true);
```
Comment (turtledreams) [inline Assets/Tests/PlayModeTests/DeviceIdTests.cs:398]: I would like early return and removing else, for better readability or you can change the device id you feed to validate function with if check and have one validatedeviceIDandtype only. You think the same?

File: Assets/Tests/PlayModeTests/DeviceIdTests.cs [hunk 2]
```diff
@@ -377,22 +379,100 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
...
+        // We provide the same device id that SDK generated 
+        // Id should remain same and type shouldn't change
+        [TestCase(DeviceIdType.SDKGenerated)]
+        [TestCase(DeviceIdType.DeveloperProvided)]
+        public void SetId_SameId(DeviceIdType startType)
+        {
+            string deviceId = "";
+            if (startType == DeviceIdType.SDKGenerated)
+            {
+                CountlyConfiguration config = new CountlyConfiguration(_appKey, _serverUrl);
```
Comment (turtledreams) [inline Assets/Tests/PlayModeTests/DeviceIdTests.cs:441]: here again set the config and device id type alone to reduce repetition

File: Assets/Tests/PlayModeTests/DeviceIdTests.cs [hunk 3]
```diff
@@ -377,22 +379,100 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
...
+            {
+                Countly.Instance.Init(TestUtility.CreateBaseConfig());
+                deviceId = Countly.Instance.Device.DeviceId; // "test_user"
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.DeveloperProvided, true);
+
+                Countly.Instance.Device.SetId(deviceId);
+                ValidateDeviceIDAndType(Countly.Instance, deviceId, DeviceIdType.DeveloperProvided, true);
+            }
+        }
+
```
Comment (turtledreams) [inline Assets/Tests/PlayModeTests/DeviceIdTests.cs:460]: can you add 1 more test where you double call setid with different values and verify at each step id is correct

**Review by turtledreams — CHANGES_REQUESTED** [review]

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
styling needs to be fixed

File: Assets/Tests/PlayModeTests/DeviceIdTests.cs [hunk 4]
```diff
@@ -377,6 +372,76 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
...
+        // We provide empty, null, whitespace and valid ids to SetId method over SDKGenerated id
+        // Id should only change in case of a valid id
+        [TestCase("", DeviceIdType.SDKGenerated, false)]
+        [TestCase(null, DeviceIdType.SDKGenerated, false)]
+        [TestCase(" ", DeviceIdType.SDKGenerated, false)]
+        [TestCase("new_user", DeviceIdType.DeveloperProvided, true)]
+        public void SetId_OverSDKGenerated(string id, DeviceIdType expectedIdType, bool isExpectingChange)
+        {
+            CountlyConfiguration config = new CountlyConfiguration(_appKey, _serverUrl);
+            Countly.Instance.Init(config);
```
Comment (ijunaid) [inline Assets/Tests/PlayModeTests/DeviceIdTests.cs:406]: Why aren't we just using TestUtility.CreateBaseConfig() instead of making a whole new config?

**Review by ijunaid — APPROVED** [review]
Overall looks good to me, some if else conditions can improve to remove the repeated code as D mentioned.

File: Assets/Countly/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs [hunk 5]
```diff
@@ -174,6 +184,32 @@ public async Task ChangeDeviceIdWithMerge(string deviceId)
             await Task.CompletedTask;
         }
 
+        private async void SetIdInternal(string newDeviceId)
```
Comment (turtledreams) [inline Assets/Countly/Plugins/CountlySDK/Services/DeviceIdCountlyService.cs:187]: seems like this should be Task instead of void <3

**Review by turtledreams — CHANGES_REQUESTED** [review]

File: Assets/Tests/PlayModeTests/DeviceIdTests.cs [hunk 6]
```diff
@@ -377,6 +372,84 @@ public void GeneratedDeviceID_CustomDeviceIDProvided()
...
+        // 'SetId' method in DeviceIdCountlyService
+        // We provide the same device id that SDK generated 
+        // Id should remain same and type shouldn't change
+        [TestCase(DeviceIdType.SDKGenerated)]
+        [TestCase(DeviceIdType.DeveloperProvided)]
+        public void SetId_SameId(DeviceIdType startType)
+        {
+            string deviceId;
+            DeviceIdType currentType;
+            if (startType == DeviceIdType.SDKGenerated) {
```
Comment (turtledreams) [inline Assets/Tests/PlayModeTests/DeviceIdTests.cs:414]: this code is perfectly fine. a suggestion though: you can assign a CountlyConfiguration to a configuration from the if block or the else block initially. then with a single if check you can re-assign it if necessary. this way you would eliminate the else block. only if that would be sth you also like

**Review by turtledreams — CHANGES_REQUESTED** [review]
Yeap good improvements. I think we are almost there. Only thing remains is to check if we are merging or not merging.
so we check type and id before and after setID. But we do not check if that change happened with merge or without merge (or it seems). might use same validators that exist for change id methods if they exist

**Review by turtledreams — CHANGES_REQUESTED** [review]
merge conflict

**Comment by AliRKat** [comment]
merge conflict is solved

**Review by turtledreams — APPROVED** [review]

**Review by ijunaid — APPROVED** [review]

**Pull request merged**